CONTRIBUTING.md

```diff
@@ -550,6 +550,239 @@ cat generated/watcher/watcher.log  # See k_ids field for each core in the last d
+with the branch.
+
+#### Creating the initial Pull Request (PR)
+
+You need to have an issue. Either someone creates one and assigns it to you, or
+you need to create an issue. For example issue 123.
+
+#### Creating a branch on a VM
+
+Include the user, the issue number, and optionally a description of the change.
+/ and - are used as separators between user and issue number. And - and _
+between issue number and description. E.g.
+
+```
+git checkout -b user-123
+git checkout -b user/123
+git checkout -b user-123_rename_method_x
+git checkout -b user/123-add-x-unit-test
+```
+
+#### Saving your changes
+
+Edit the files that you want, making sure relevant unit tests pass. Then add
+them in. E.g.
+
+```
+git add abc.py
+git add -A
+```
+
+You can restore files if you need to get the original. E.g.
+
+```
+git restore abc.py
+git restore '*'
+git restore --staged abc.py # if the file was already added
+```
+
+Once you are satisfied that everything works, create a commit. The message
+needs to specify the issue number, with a pound sign and a colon, and the
+description should describe the change itself, and cannot be generic and refer
+to something else, such as saying that you are addressing reviewer feedback.
+
+```
+git commit -m"#123: rename method x"
+```
+
+Note: each commit should work. That is, everything compiles properly on the
+architecture used by your VM, you can run relevant code on the card, and
+relevant unit tests pass.
+
+You can use git log to see the sequence of commits in the branch. That allows
+you to see where your branch is relative to main, and can help you figure out
+how the commits are structured, before and after commits and rebases.
+
+#### Saving the commit to origin and create a pull request
+
+You will need to push the change to origin. The command will provide a url that
+you should use to create pull request. This should be done the first time you
+push a change. After that you may need to set upstream to be able to push
+changes in the future. E.g.
+
+```
+git push origin user-123:user-123
+git branch --set-upstream-to=origin/user-123 user-123
+```
+
+Note: you may be able to push and set the upstream at the same time, but that
+assumes that you haven't rebased, which is probably not the case. The command
+would be something like
+
+```
+git push origin --set-upstream origin user-123
+```
+
+If that doesn't work, you should use `branch --set-upstream-to`.
+
+Once you have a pull request, in the UI you can run actions against the branch.
+Go to Actions (https://github.com/tenstorrent/tt-metal/actions) and run the
+workflows that you want against your branch. At the very least, you should run
+All post-commit tests.
+
+You can make more changes, commit them, and then if everything is set up and you
+don't need to rebase, then you can just do
+
+```
+git push
+```
+
+Occasionally, and for the final set of tests before the final commit, you should
+rebase your branch.
+
+#### Rebasing your branch
```

Comment:
Number 1 thing is to make sure that git configured correctly. Default behavior is that `#` character is used for comments, so our `#issue-number: commit description` pattern breaks rebasing till default comment character is changed to something else.

Reply:
This info was present in our setup docs, but I recommend to point this here.

Reply:
Added a comment. This is discussed earlier in the doc.